Proposal to add counting options

emugel opened the pull request:
Following #40, this is a proposal adding options to how the current count is shown in the below screen:

```
  +---------------------+
  |                  [:]|
  |    Cigarets         |
  |                     |
  |                     |
  | 🕖 20/06/20 13:09   |
  | ------------------  |
  | 153 Tracked     [+] |
  +---------------------+
```

In short the purpose is to show a count for today, this week or this month, while also working with `Multiple choice` type of data, and the potential future time ranges proposed in #15.

The default would be to show the total count in the data series (as it is now), but user could choose to display it differently for each data series, for instance below using "Total today":

```
  +---------------------+
  |                  [:]|
  |    Cigarets         |
  |                     |
  |                     |
  | 🕖 20/06/20 13:09   |
  | ------------------  |
  | 3 Today         [+] |
  +---------------------+

  +---------------------+
  |                  [:]|
  |    Drinking         |
  |                     |
  |                     |
  | 🕖 19/06/20 22:39   |
  | ------------------  |
  | 5 Today         [+] |
  +---------------------+

```

This is only a text mockup for now, as it uses the same components as before (label, checkbox, pulldown menu):

```
What would you like to track?
_Cigarets___________

What type of data is this?

  Numerical v
  Multiple

 [x] Set a default value

 _____________1.0___________

 [x] Show a special count

  Tracked (default)  v
  Total Today
  Total this week
  Total this month

```

When the `[ ] Show a special count` checkbox is off, the pulldown menu below does not appear.

- For **numerical**, it can be used to show total, daily, weekly or monthly count.
- For **multiple choice**, there seems to be 2 distinct cases:

  1. index doesn't matter (e.g. a `{blue=0, red=1, yellow=2}` partition): in that case the user seems to have no reason to want to change from "Tracked (Default)" to another value. It would work, but giving e.g. a total for the week has no reason to be useful in that case.
  2. index has a significant value (e.g. drinking ∈ `{none=0, light=1, heavy=2}`), in that case the following values seem useful:
      - *Tracked (Default)* (as it is now)
      - *Total today*
      - *Total this week*
      - *Total this month*


# More

If idea **start and stop tracking** (#15), gets implemented, we could use this value to show the total amount of time ranges (Total, daily, weekly and monthly may also work). 

In case of a running time range (I mean when the chronometer is "running") it would mean the total value shown would have to be updated in real time. If is a problem we can decide to not have the checkbox `Show a special count` clickable when the data type is not numerical or multiple choice. 

Lastly, a general setting to offset the time for day changing (the value to use for midnight) can be imagined. But it seems preferable to leave it for another proposal, as we don't know yet if it is necessary.

emugel commented:
(pls wait some 10min need to edit from my shitty connection...)

SamAmco commented:
@emugel Thank you for taking an interest and putting together this proposal, I appreciate the effort. 

Reading your proposal a couple of issues stand out for me and I think I may have a preferable solution I would like to hear your thoughts on.

1. First the most common criticism I get with Track & Graph is that it is hard to understand at first. I think a lot of this stems from the fact that the first thing you need to do is go 2 menu's deep and start configuring something before you're really sure what the outcome of this configuration will be. So I am always weary of adding more complexity to the feature setup screen (AddFeatureFragment.kt). I think there probably needs to be slightly more "hand holding" for first time users and this becomes hard to do if you have configuration options that are hard to explain to someone who has never seen the app before. 

2. I don't think it is necessarily a given that those tracking multiple choice data where "index doesn't matter" would not want to take advantage of this feature. For example say `{blue=0, red=1, yellow=2}` represents the colours of birds you saw at the park. You may want to draw a pie chart showing the ratios of birds seen, but you may also be interested in how many birds were seen in total. So I can imagine you might want to break this into two configuration options as outlined below.

### Counter proposal:

There is currently a popup menu button in the top right of feature cards:

![Screenshot_20200620_124953_com samco trackandgraph debug](https://user-images.githubusercontent.com/5012146/85201247-c7615d80-b2f5-11ea-88fe-a01c26b4ad18.jpg)

I propose we add an item to this menu named "Customise". This option would bring up a dialog fragment (see the "Description" or "Move to" options for an example) that would allow you to customise the look and feel of the card UI. There may be more customisation we want to offer in future, but for now I think it would be something like:

![Screenshot from 2020-06-20 13-22-10](https://user-images.githubusercontent.com/5012146/85201679-473cf700-b2f9-11ea-8e7f-d5cbded18331.png)

Where "Total by:" would have the options `{"Number of data points", "Sum of data points"}` and "Total for:" would have the options `{"All data", "This Year", "This Month", "This Week", "Today"}`

The way I see it there may be value in drawing a distinction between the feature configuration (its name, description, type...) and the feature card configuration. In future it may be the case that it makes more sense to configure default values in the customise dialog instead of the add feature fragment since this again is purely to do with how the card its self behaves, but I suggest we leave this as is for now. 

From a database perspective I don't think there's much point in making this distinction yet, I would suggest that you add two new enums to Feature.kt for "card_total_by" and "card_total_for" (or something like that). Don't forget you will then need to write type converters for these enums (see TrackAndGraphDatabase.kt) and you will need to write a migration (see DatabaseMigrations.kt).

I suggest you hold off for a short while as I am currently partially implementing https://github.com/SamAmco/track-and-graph/issues/15 by adding time durations as a feature type and we are likely to get significant merge conflicts. I will let you know here when I've merged this, it shouldn't be more than a day or two.

emugel commented:
OK I wait.
will respond in more details when I have more time, 
but I mostly agree :)

SamAmco commented:
Hi @emugel I have now merged all branches to master and created a new release candidate rc-v1.9.0. I suspect you're fine to start working from here without too much conflict in the near future. I am planning more work in the soon but it is mostly on the graphing side of the app.

emugel commented:
Ok! Note I may not be starting right away, as I never even written in Kotlin before
(I normally use Haxe, but these languages are supposed to be quite similar).

So think I will make an attempt somewhere in the month of July.
Will probably have questions then, will keep you posted :)

SamAmco commented:
No worries, I will try to help where I can. I am on furlough at the moment so I am working on Track & Graph pretty much full time but I am likely to be back working full time by about the 20th at which point I will likely have much less time to commit.

emugel commented:
Hum I am not very sure I can do it anymore. I managed to add context menu "Customise card" (showing an alert box) adding enums and their fields in `database.element.Feature.kt`, and dealt with the migration. But this was not conceptually difficult. 

Now I'm facing problems with the dialog itself and I'm bit lost. One the one hand it seems like Description dialog is too simple and MoveTo dialog is too complex.

Amongst questions I have are those:
- I don't know if my view needs a viewModel,
- I am not sure what these "discrete" things are and if my 2 simple enums need them,
- and I don't even know if I can use an alert window :)

Anyway I'm going to post this message now, I'm off for a day or two.

emugel commented:
I decided to follow more closely the way the MoveTo dialog is written as opposed to the Description.
Will see how it goes ;)

emugel commented:
Some progress done on https://github.com/emugel/track-and-graph/pull/new/implement_proposal_pr41

* added a "Customise card" menu and a dialog with 2 spinners
* 2 new fields and enums added to Feature.kt
* added migration and enum converters

The spinners in the dialog already change the model persistently.
Remains to effect the actual counters.

(currently there is nothing done about the new "time period" feature that I just noticed a few hours ago :)
